docs/dev/kube-apiserver-health-check.md

```diff
@@ -0,0 +1,70 @@
+## Graceful Termination
+`kube-apiserver` in OpenShift is fronted by an external and an internal Load Balancer. This document serves as a
+guideline on how to properly configure the health check probe of the load balancer so that when a `kube-apiserver`
+instance restarts we can ensure:
+- The load balancer detects it and takes it out of service in time. No new request should be forwarded to the 
+  `kube-apiserver` instance when it has stopped listening.
+- Existing connections are not cut off hard, they are allowed to complete gracefully.
+
+## Load Balancer Health Check Probe
+`kube-apiserver` provides graceful termination support via the `/readyz` health check endpoint. When `/readyz`reports
+`ok` it indicates that the apiserver is ready to serve request(s).
+
+Now let's walk through the events (in chronological order) that unfold when a `kube-apiserver` instance restarts:
+* E1: `T+0s`: `kube-apiserver` receives a TERM signal.
+* E2: `T+0s`: `/readyz` starts reporting `failure` to signal to the load balancer that a shut down is in progress.
+  * The apiserver will continue to accept new request(s).
+  * The apiserver waits for certain amount of time (configurable by `shutdown-delay-duration`) before it stops accepting new request(s).
+* E3: `T+70s`: `kube-apiserver` (the http server) stops listening:
+  * `/healthz` turns red.
+  * Default TCP health check probe on port `6443` will fail.
+  * Any new request forwarded to it will fail, most likely with a `connection refused` error.
+  * Existing request(s) in-flight are not cut off but are allowed to complete gracefully, but there is a timeout of `60s`.
+* E4: `T+70s+60s`: Any existing request(s) that are still in flight are terminated.
+* E5: `T+70s+60s`: The apiserver process exits.
```

Comment:
these last two steps can happen between T+70s and T+70s+60s, depending on in-flight requests.

Reply:
fixed